### ISSUE TITLE
Java 9 support

### ISSUE THREAD

**Issue by dariokures** [glowroot/glowroot#300]
Hi,

Is Java 9 supported by Glowroot agent?
I've tried it with Tomcat 9.0.2 and JDK 9.0.1, but I'm getting this error:
```
ERROR o.glowroot.agent.weaving.AdviceCache - error creating advice: org.glowroot.agent.plugin.jdbc.DataSourceAspect$GetConnectionAdvice
java.lang.NoClassDefFoundError: java/sql/Connection
        at java.base/java.lang.Class.getDeclaredMethods0(Native Method)
        at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3139)
        at java.base/java.lang.Class.privateGetPublicMethods(Class.java:3164)
        at java.base/java.lang.Class.getMethods(Class.java:1861)
        at org.glowroot.agent.weaving.AdviceBuilder.build(AdviceBuilder.java:155)
        at org.glowroot.agent.weaving.AdviceCache.getAdvisors(AdviceCache.java:171)
        at org.glowroot.agent.weaving.AdviceCache.<init>(AdviceCache.java:89)
        at org.glowroot.agent.init.AgentModule.<init>(AgentModule.java:134)
        at org.glowroot.agent.embedded.init.EmbeddedAgentModule.<init>(EmbeddedAgentModule.java:139)
        at org.glowroot.agent.embedded.init.EmbeddedGlowrootAgentInit.init(EmbeddedGlowrootAgentInit.java:59)
        at org.glowroot.agent.MainEntryPoint.start(MainEntryPoint.java:195)
        at org.glowroot.agent.MainEntryPoint.premain(MainEntryPoint.java:111)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at org.glowroot.agent.AgentPremain.premain(AgentPremain.java:54)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:500)
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:512)
WARN  o.g.agent.weaving.ClassLoaders - sun/misc/Launcher
java.lang.ClassNotFoundException: sun/misc/Launcher
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:375)
        at org.glowroot.agent.weaving.ClassLoaders.appendToBootstrapResourcePath(ClassLoaders.java:146)
        at org.glowroot.agent.weaving.ClassLoaders.defineClassesInBootstrapClassLoader(ClassLoaders.java:69)
        at org.glowroot.agent.weaving.AdviceCache.<init>(AdviceCache.java:117)
        at org.glowroot.agent.init.AgentModule.<init>(AgentModule.java:134)
        at org.glowroot.agent.embedded.init.EmbeddedAgentModule.<init>(EmbeddedAgentModule.java:139)
        at org.glowroot.agent.embedded.init.EmbeddedGlowrootAgentInit.init(EmbeddedGlowrootAgentInit.java:59)
        at org.glowroot.agent.MainEntryPoint.start(MainEntryPoint.java:195)
        at org.glowroot.agent.MainEntryPoint.premain(MainEntryPoint.java:111)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at org.glowroot.agent.AgentPremain.premain(AgentPremain.java:54)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:500)
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:512)
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.glowroot.agent.init.GaugeCollector$1 to method sun.management.ManagementFactoryHelper.registerInternalMBeans(javax.management.MBeanServer)
WARNING: Please consider reporting this to the maintainers of org.glowroot.agent.init.GaugeCollector$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
INFO  org.glowroot - plugins loaded: Cassandra Plugin, Elasticsearch Plugin, Executor Plugin, Grails Plugin, Hibernate Plugin, HTTP Client Plugin, Java HTTP Server Plugin, JAX-RS Plugin, Jdbc Plugin, JMS Plugin, JSFPlugin, JSP Plugin, Logger Plugin, Play Plugin, Netty Plugin, Quartz Plugin, Redis Plugin, Servlet Plugin, Spring Plugin, Struts Plugin
ERROR org.glowroot - Glowroot not started: java/sql/Array
java.lang.NoClassDefFoundError: java/sql/Array
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:375)
        at org.glowroot.agent.embedded.init.PreInitializeStorageShutdownClasses.initialize(PreInitializeStorageShutdownClasses.java:55)
        at org.glowroot.agent.embedded.init.PreInitializeStorageShutdownClasses.preInitializeClasses(PreInitializeStorageShutdownClasses.java:39)
        at org.glowroot.agent.embedded.init.EmbeddedAgentModule.<init>(EmbeddedAgentModule.java:144)
        at org.glowroot.agent.embedded.init.EmbeddedGlowrootAgentInit.init(EmbeddedGlowrootAgentInit.java:59)
        at org.glowroot.agent.MainEntryPoint.start(MainEntryPoint.java:195)
        at org.glowroot.agent.MainEntryPoint.premain(MainEntryPoint.java:111)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at org.glowroot.agent.AgentPremain.premain(AgentPremain.java:54)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:500)
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:512)
```

Thanks,
Dario

**Comment by trask** [comment]
Hi @dariokures, thanks for requesting this! I've been working on Java 9 support, and it will hopefully be available in the next release.

**Comment by trask** [comment]
The latest snapshot release [glowroot-agent-0.10.0-SNAPSHOT-dist.zip](https://oss.sonatype.org/service/local/artifact/maven/redirect?r=snapshots&g=org.glowroot&a=glowroot-agent&c=dist&e=zip&v=0.10.0-SNAPSHOT) should run under Java 9 now when reporting to a central collector, but doesn't work yet when running with local UI/storage.

**Comment by dariokures** [comment]
Tested with some rest web services(jax-rs, jdbc), everything looks good. Thanks!
Note that default /examples context on tomcat 9.0.2 doesn't work with glowroot agent:
```
SEVERE [main] org.apache.catalina.core.StandardContext.listenerStart Exception sending context initialized event to listener instance of class [websocket.drawboard.DrawboardContextListener]
 java.lang.NoSuchMethodError: java.lang.Object.contextInitialized(Ljavax/servlet/ServletContextEvent;)V
	at websocket.drawboard.DrawboardContextListener.contextInitialized(DrawboardContextListener.java)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4643)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5105)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:740)
	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:716)
	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:703)
	at org.apache.catalina.startup.HostConfig.deployDirectory(HostConfig.java:1131)
	at org.apache.catalina.startup.HostConfig$DeployDirectory.run(HostConfig.java:1865)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:514)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)
	at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:118)
	at org.apache.catalina.startup.HostConfig.deployDirectories(HostConfig.java:1043)
	at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:433)
	at org.apache.catalina.startup.HostConfig.start(HostConfig.java:1575)
	at org.apache.catalina.startup.HostConfig.lifecycleEvent(HostConfig.java:308)
	at org.apache.catalina.util.LifecycleBase.fireLifecycleEvent(LifecycleBase.java:123)
	at org.apache.catalina.util.LifecycleBase.setStateInternal(LifecycleBase.java:424)
	at org.apache.catalina.util.LifecycleBase.setState(LifecycleBase.java:367)
	at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:964)
	at org.apache.catalina.core.StandardHost.startInternal(StandardHost.java:839)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1425)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1415)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)
	at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:140)
	at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:941)
	at org.apache.catalina.core.StandardEngine.startInternal(StandardEngine.java:258)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.core.StandardService.startInternal(StandardService.java:422)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.core.StandardServer.startInternal(StandardServer.java:770)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.startup.Catalina.start(Catalina.java:671)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.apache.catalina.startup.Bootstrap.start(Bootstrap.java:353)
	at org.apache.catalina.startup.Bootstrap.main(Bootstrap.java:493)
```

DrawboardContextListener.java on tomcat 8.5.24 (ok)
```java
package websocket.drawboard;

import javax.servlet.ServletContextEvent;
import javax.servlet.ServletContextListener;

public final class DrawboardContextListener implements ServletContextListener {

    @Override
    public void contextInitialized(ServletContextEvent sce) {
        // NO-OP
    }

    @Override
    public void contextDestroyed(ServletContextEvent sce) {
        // Shutdown our room.
        Room room = DrawboardEndpoint.getRoom(false);
        if (room != null) {
            room.shutdown();
        }
    }
}
```

DrawboardContextListener.java on tomcat 9.0.2 (doesn't work with glowroot agent)
```java
package websocket.drawboard;

import javax.servlet.ServletContextEvent;
import javax.servlet.ServletContextListener;

public final class DrawboardContextListener implements ServletContextListener {

    @Override
    public void contextDestroyed(ServletContextEvent sce) {
        // Shutdown our room.
        Room room = DrawboardEndpoint.getRoom(false);
        if (room != null) {
            room.shutdown();
        }
    }
}
```

**Comment by trask** [comment]
Hi @dariokures, thanks for reporting this! I'm able to reproduce this and will look into it.

**Comment by trask** [comment]
Hi @dariokures, the tomcat 9 examples issue turned out to be an issue with weaving of Java 8 default methods. Thanks again for reporting it. It should be fixed now by 4ad149a38908fd9662e47e6283d1ae004218ac1a in the latest snapshot release [glowroot-agent-0.10.0-SNAPSHOT-dist.zip](https://oss.sonatype.org/service/local/artifact/maven/redirect?r=snapshots&g=org.glowroot&a=glowroot-agent&c=dist&e=zip&v=0.10.0-SNAPSHOT).

**Comment by trask** [comment]
Java 9 support is available in the latest snapshot release [glowroot-agent-0.10.3-SNAPSHOT-dist.zip](https://oss.sonatype.org/service/local/artifact/maven/redirect?r=snapshots&g=org.glowroot&a=glowroot-agent&c=dist&e=zip&v=0.10.3-SNAPSHOT).  Please report any issues.
